.taskcluster.yml

```diff
@@ -2,6 +2,334 @@ version: 1
+                                  - '--task-cwd=/builds/worker/checkouts/src'
+                                  - '--'
+                                  - bash
+                                  - -cx
+                                  - $let:
+                                        extraArgs: {$if: 'tasks_for == "cron"', then: '${cron.quoted_args}', else: ''}
+                                    in:
+                                        $if: 'tasks_for == "action"'
+                                        then: >
+                                            PIP_IGNORE_INSTALLED=0 pip3 install --user /builds/worker/checkouts/taskgraph &&
+                                            PIP_IGNORE_INSTALLED=0 pip3 install --user mozilla-version &&
+                                            taskcluster/scripts/decision-install-sdk.sh &&
+                                            ln -s /builds/worker/artifacts artifacts &&
+                                            ~/.local/bin/taskgraph action-callback
+                                        else: >
+                                            PIP_IGNORE_INSTALLED=0 pip3 install --user /builds/worker/checkouts/taskgraph &&
+                                            PIP_IGNORE_INSTALLED=0 pip3 install --user mozilla-version &&
+                                            taskcluster/scripts/decision-install-sdk.sh &&
+                                            ln -s /builds/worker/artifacts artifacts &&
+                                            ~/.local/bin/taskgraph decision
+                                            --pushlog-id='0'
+                                            --pushdate='0'
+                                            --project='${project}'
+                                            --message=""
+                                            --owner='${ownerEmail}'
+                                            --level='${level}'
+                                            --base-repository="$MOBILE_BASE_REPOSITORY"
+                                            --head-repository="$MOBILE_HEAD_REPOSITORY"
+                                            --head-ref="$MOBILE_HEAD_REF"
+                                            --head-rev="$MOBILE_HEAD_REV"
+                                            --head-tag="$MOBILE_HEAD_TAG"
+                                            --repository-type="$MOBILE_REPOSITORY_TYPE"
+                                            --tasks-for='${tasks_for}'
+                                            ${extraArgs}
+
+                              artifacts:
+                                  'public':
+                                      type: 'directory'
+                                      path: '/builds/worker/artifacts'
+                                      expires: {$fromNow: '1 year'}
+                                  'public/docker-contexts':
```

Comment:
I'm seeing an error in the decision task at the end of the log stating:
```
[task 2021-08-17T15:42:48.301Z] bash: taskcluster/scripts/decision-install-sdk.sh: No such file or directory
[taskcluster 2021-08-17 15:42:48.532Z] === Task Finished ===
[taskcluster 2021-08-17 15:42:48.640Z] Artifact "public/docker-contexts" not found at "/builds/worker/checkouts/src/docker-contexts"
```

This was copied over from Fenix so is this specific to that repository?

Reply:
On line 367 (from commented out config), I see this:
```
artifacts:
  #         public:
  #           type: directory
  #           path: /opt/focus-android/app/build/reports
  #           expires: {$fromNow: '1 week'}
```
Should I be using that path instead for artifacts? @bhearsum please advise.

Reply:
I think the real error here is that:
`bash: taskcluster/scripts/decision-install-sdk.sh: No such file or directory`

If I had to guess, this is causing the Decision task to terminate early (but without raising an exception?) before it gets to the point of writing out artifacts.. Then taskcluster notices that we declared an artifact that doesn't exist and complains about it.

I could be wrong though..

Reply:
That sounds right to me. You'll need to copy in https://github.com/mozilla-mobile/fenix/blob/main/taskcluster/scripts/decision-install-sdk.sh to fix it.

Reply:
Two things I noticed: The sdk installation is baked into the dockerfile here: https://github.com/mozilla-mobile/focus-android/pull/5145/files#diff-50d0c8ed55e4948f99160e4e1da1d5bedb8b620b289b724a9eb977fe1e13c458R42

Is the idea that this will be extracted into a separate file in that script instead?

Also, @bhearsum I saw in your VPN pr that you don't install this script in the taskcluster.yml else statement the way fenix does (and what I copied into focus): https://github.com/mozilla-mobile/mozilla-vpn-client/pull/1562/files#diff-a728f7e52d751b98eafa856e45594533339b44f229d7b83f930df335391e7e15R266

I'm curious why, and which is the correct way for Focus? :)

Reply:
The Dockerfile you've linked to there is the base image for build, test, etc. tasks. The issue you're hitting here is with _decision_ tasks, which run using the Docker image specified at https://github.com/mozilla-mobile/focus-android/pull/5145/files#diff-a728f7e52d751b98eafa856e45594533339b44f229d7b83f930df335391e7e15R240 -- which is why we need to install the SDK there as well. (Decision tasks don't use the same images for reasons I can't totally recall at the moment...)

> Also, @bhearsum I saw in your VPN pr that you don't install this script in the taskcluster.yml else statement the way fenix does (and what I copied into focus): https://github.com/mozilla-mobile/mozilla-vpn-client/pull/1562/files#diff-a728f7e52d751b98eafa856e45594533339b44f229d7b83f930df335391e7e15R266

The reason it's not needed for VPN is because there's nothing the decision task is doing that needs the Android SDK. It may, in the future, when we're doing the Android builds in Taskcluster, but its build system is also significantly different than the browser products, so it's unclear.

Focus's (and the other browsers') build & task scheduling is deeply intertwined with Android and Gradle, so they need both available during the decision task.

Reply:
> The Dockerfile you've linked to there is the base image for build, test, etc. tasks. The issue you're hitting here is with _decision_ tasks, which run using the Docker image specified at https://github.com/mozilla-mobile/focus-android/pull/5145/files#diff-a728f7e52d751b98eafa856e45594533339b44f229d7b83f930df335391e7e15R240 -- which is why we need to install the SDK there as well. (Decision tasks don't use the same images for reasons I can't totally recall at the moment...)
> 

hrm, so I moved that dockerfile (that's the only one they have) into `taskcluster/docker/base`. Not sure what file you were linking to in the diff. Was it this? https://github.com/mozilla-mobile/focus-android/pull/5145/files#diff-a1854fd13c43b022659a78b2e7a76c978bcada7d3795c30463ac250520161666R31

> The reason it's not needed for VPN is because there's nothing the decision task is doing that needs the Android SDK. It may, in the future, when we're doing the Android builds in Taskcluster, but its build system is also significantly different than the browser products, so it's unclear.
> 
> Focus's (and the other browsers') build & task scheduling is deeply intertwined with Android and Gradle, so they need both available during the decision task.

I think that makes sense, but this still seemed to work by removing the sdk installation in the else block.

Reply:
> hrm, so I moved that dockerfile (that's the only one they have) into taskcluster/docker/base. Not sure what file you were linking to in the diff. Was it this? https://github.com/mozilla-mobile/focus-android/pull/5145/files#diff-a1854fd13c43b022659a78b2e7a76c978bcada7d3795c30463ac250520161666R31

I was trying to link to `.taskcluster.yml` actually :). There's a block in there that specifies the docker image to use for the decision task. Let's try this link: https://github.com/mozilla-mobile/focus-android/blob/09d21aaa7e3727b0b540182e052bc81ffd56bfaa/.taskcluster.yml#L239